keras_nlp/layers/transformer_decoder.py

```diff
@@ -0,0 +1,226 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transformer decoder block implementation based on `keras.layers.Layer`."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.layers.transformer_layer_utils import (  # isort:skip
+    compute_causal_mask,
+    merge_padding_and_attention_mask,
+)
+
+
+class TransformerDecoder(keras.layers.Layer):
+    """Transformer decoder.
+
+    This class follows the architecture of transformer decoder layer in paper
+    "Attention is All You Need" (https://arxiv.org/abs/1706.03762). Users can
+    instantiate multiple instances of this class to stack up the decoder.
+
+    Args:
+        intermediate_dim: int, the hidden size of feedforward network.
+        num_heads: int, the number of heads in MultiHeadAttention.
+        dropout: float, defaults to 0. the dropout value, shared by
+            MultiHeadAttention and feedforward network.
+        activation: string or tf.keras.activations, defaults to "relu". the
+            activation function of feedforward network.
+        layer_norm_epsilon: float, defaults to 1e-5. The eps value in layer
+            normalization components.
+        name: string, defaults to "TransformerDecoder". The name of the layer.
+        **kwargs: other keyword arguments.
+
+    Examples:
+    ```python
+    # Create a single transformer decoder layer.
+    decoder = keras_nlp.layer.TransformerDecoder(
+        intermediate_dim=64, num_heads=8)
+
+    # Create a simple model containing the decoder.
+    decoder_input = tf.keras.Input(shape=[4, 6])
+    encoder_input = tf.keras.Input(shape=[4, 6])
+    output = decoder(decoder_input, encoder_input)
+    model = tf.keras.Model(inputs=[decoder_input, encoder_input],
+        outputs=output)
+
+    # Call decoder on the inputs.
+    decoder_input_data = tf.random.uniform(shape=[1, 10, 64])
+    encoder_input_data = tf.random.uniform(shape=[1, 10, 64])
+    decoder_output = model([decoder_input_data, encoder_input_data])
+
+    ```
+
+    References:
+        [Vaswani et al., 20XX](https://arxiv.org/abs/1706.03762)
+
+    """
+
+    def __init__(
+        self,
+        intermediate_dim,
+        num_heads,
+        dropout=0,
+        activation="relu",
+        layer_norm_epsilon=1e-05,
+        name="TransformerDecoder",
```

Comment:
Names should default to None and should never be hardcoded in the signature. Otherwise multiple layers are going to end up with the same name. The base Layer class takes care of default name autogeneration.

Reply:
Done.